Archaeologist Background (SteelNeuron)

Vajrapani opened the pull request:
Thanks to SteelNeuron, and also bhauth

PabloMansanet committed:
Archaeologist shell

PabloMansanet committed:
Initial stats and appearance

PabloMansanet committed:
Working archaeologist background

PabloMansanet committed:
Reduce manual skill points

PabloMansanet committed:
Manual and randart tweaks

PabloMansanet committed:
Support forgetful archaeologists that dropped their gear somewhere

Commit:
archaeologist conflict resolution

Commit:
fix parens

Commit:
whitespace

Vajrapani committed:
Fix some old unrands & add in an 'include'

To make it compile successfully

Vajrapani committed:
Give 3 Stealth / 3 Dodging / 1 Fighting

Slightly better starting skills

NormalPerson7 commented on crawl-ref/source/evoke.cc:
```diff
@@ -520,13 +522,58 @@ bool skill_has_manual(skill_type skill)
+        int crate_index = -1;
+        for (int i = 0; i < ENDOFPACK; i++)
+            if (you.inv[i].defined() && you.inv[i].is_type(OBJ_MISCELLANY, MISC_ANCIENT_CRATE))
+                crate_index = i;
```
These two if statements are too long, so do them as:
`if (a`
`----&& b)`

NormalPerson7 commented on crawl-ref/source/evoke.cc:
```diff
@@ -520,13 +522,58 @@ bool skill_has_manual(skill_type skill)
+    item_colour(tome);
+    item_set_appearance(tome);
+    mprf("You have an epiphany! The dusty tome is %s! Reading it may be key to unlocking the crate...",
+         tome.name(DESC_A).c_str());
```
This line is also too long.

NormalPerson7 commented on crawl-ref/source/items.cc:
```diff
@@ -1801,6 +1802,12 @@ bool move_item_to_inv(int obj, int quant_got, bool quiet)
     const coord_def old_item_pos = it.pos;
 
+    if (you.props.exists(ARCHAEOLOGIST_TRIGGER_TOME_ON_PICKUP) && it.props.exists(ARCHAEOLOGIST_TOME_SKILL))
+        archaeologist_read_tome(it);
```
Again do these ifs as
`if (a`
`----&& b)`

NormalPerson7 commented on crawl-ref/source/job-data.h:
```diff
@@ -54,6 +54,15 @@ static const map<job_type, job_def> job_data =
+    3, 6, 3,
+    { SP_HUMAN, SP_TENGU, SP_DEEP_ELF, SP_DEEP_DWARF, SP_DEMONSPAWN},
+    { "robe plus:1", "hat plus:1", "pair of boots plus:1", "pair of gloves plus:1", "dusty tome", "ancient crate" },
+    WCHOICE_NONE,
```
Another line too long...

NormalPerson7 commented on crawl-ref/source/ng-setup.cc:
```diff
@@ -235,6 +237,167 @@ static void _give_ammo(weapon_type weapon, int plus)
+    UNRAND_TALOS,                       // armour of Talos
+    UNRAND_WARLOCK_MIRROR,              // warlock's mirror
+};
+
```
Ermm.... can't you just use art-enum.h and checking them for the nogen flag rather than listing them all out? This seems entirely unnecessary, and if you want to forbid certain unrands from being generated here, you should add a new flag.

NormalPerson7 commented on crawl-ref/source/ng-setup.cc:
```diff
@@ -235,6 +237,167 @@ static void _give_ammo(weapon_type weapon, int plus)
+            continue;
+
+    } while ((!you.could_wield(unrand) && !can_wear_armour(unrand, false, true)) ||  !can_generate(unrand));
+
```
For this `do-while` loop, you need to place the braces `{` and `}` on their own separate lines and format the `while` clause as follows for code readability (once again line too long):
`----while(!you.could_wield(unrand) && !can_wear_armour(unrand, false, true)`
`----------|| !can_generate(unrand));`
Note that the pair of brackets around the && section are not required.

NormalPerson7 commented on crawl-ref/source/ng-setup.cc:
```diff
@@ -235,6 +237,167 @@ static void _give_ammo(weapon_type weapon, int plus)
+    } while ((!you.could_wield(unrand) && !can_wear_armour(unrand, false, true)) ||  !can_generate(unrand));
+
+    dprf("Initializing archaeologist crate with %s", unrand.name(DESC_A).c_str());
+    crate.props[ARCHAEOLOGIST_CRATE_ITEM] = type;
```
dprf() line is also too long.

NormalPerson7 commented on crawl-ref/source/ng-setup.cc:
```diff
@@ -235,6 +237,167 @@ static void _give_ammo(weapon_type weapon, int plus)
+    for (uint8_t i = 0; i < ENDOFPACK; i++)
+        if (you.inv[i].defined() && you.inv[i].is_type(OBJ_MISCELLANY, MISC_ANCIENT_CRATE))
+            manual_skill = _setup_archaeologist_crate(you.inv[i]);
+
```
These ifs are again too long, so do the same as for the other ifs above.

NormalPerson7 commented on crawl-ref/source/player.cc:
```diff
@@ -3025,6 +3027,22 @@ void level_change(bool skip_attribute_increase)
+        else if(!you.props.exists(ARCHAEOLOGIST_TRIGGER_TOME_ON_PICKUP))
+            mprf("You suddenly remember the dusty tome you brought into the dungeon! You feel able"
+                 " to decipher it now. If only you could remember where you put it..."); 
+
```
More lines too long.... the if and the mprf(), and the mprf() should be an mpr().

NormalPerson7 reviewed (changes requested):
Regarding the idea: I like it.
Regarding the code: there are about fifteen instances of lines being longer than the 80 character limit which need to be fixed, and a few code style points here and there too.

There is one thing missing here: assigning JOB_ARCHAEOLOGIST some species which would prefer it: so that when you go into the background selection screen after selecting a species, Archaeologist shows up white for some species.

PabloMansanet commented on crawl-ref/source/evoke.cc:
```diff
@@ -520,13 +522,58 @@ bool skill_has_manual(skill_type skill)
+{
+    tome.base_type = OBJ_BOOKS;
+    tome.sub_type = BOOK_MANUAL;
+    tome.skill_points = 1000;
```
This should probably be 700. As it stands, I feel it unlocks a bit too late and it might be tempting to skip weaker unrands altogether and go for something else.

NormalPerson7 replied:
I'll do some maths on when this unlocks normally and see what it ends up at. Also, it will unlock earlier for low-experience-apt races I believe, which might be something I have to look into.

PabloMansanet replied:
Yes, that sounds about right :) before Lair is the target I had in mind.

Vajrapani committed:
Better formatting in 'evoke.cc'

Vajrapani committed:
Make the manual require less exp

So, the player gets their unrand faster, this incentivises them more to
build around the item they get.

From testing, the player will get this at XL 11 with 100% training

bhauth commented on crawl-ref/source/ng-setup.cc:
```diff
@@ -235,6 +237,167 @@ static void _give_ammo(weapon_type weapon, int plus)
+    switch (type)
+    {
+    case UNRAND_PONDERING:
+    case UNRAND_MAXWELL:
```
remove this line here

Vajrapani committed:
Spellcasting manual doesn't give Maxwell's

Vajrapani committed:
Better code readability

In ng-setup.cc and player.cc

NormalPerson7 commented on crawl-ref/source/ng-setup.cc:
```diff
@@ -329,15 +329,21 @@ static skill_type _setup_archaeologist_crate(item_def& crate)
 
-    } while ((!you.could_wield(unrand) && !can_wear_armour(unrand, false, true)) ||  !can_generate(unrand));
+       }
+
```
The indent on these braces isn't in-line with the `do` and the `while`. Thanks for all the other changes, they're all fine :D

Vajrapani committed:
Fix brace indents in ng-setup.cc

NormalPerson7 committed:
Use a tag and generate a vector for Ac unrands; don't use a fixed list

Previously, a fixed array `static const unrand_type _archa_unrands[]`,
defined in ng-setup.cc, was being used for generating archaeologist
unrands. This is bad for a number of maintainability reasons, most
importantly that adding a new unrand currently only requires adding an
entry to art-data.txt; with this system an entry would also need to be
added to this list.

We now add a global const function to artefact.cc (because this is the
only place unranddata is defined, and we need to use unranddata),
archaeologist_unrands(), from which archaeologist unrands can be
selected from art-enum. Additionally, it's important we use a good type
(int) for this, not some special type unrand_type which seems to be
defined only in some places.

In order to filter out a few randomly generating unrands that generate
naturally, we add a new flag notac to art-data.txt, and corresponding
code in util/art-data.pl.

There is now infrastructure to allow felids to start as an
archaeologist, however when I tried this the game crashed, so I will
keep Ac restricted for Fe for the moment. For felids only, we give
jewellery and forbid other items, and for all other races we forbid
jewellery.

Archaeologist is now almost ready to be merged.

NormalPerson7 committed:
Have Maxwell's etheric cage give a manual of Spellcasting

It would appear this was the reason for the UNRAND_MAXWELL error we had
before; UNRAND_MAXWELL is Maxwell's patent armour, and
UNRAND_ETHERIC_CAGE is Maxwell's etheric cage.

bhauth commented on crawl-ref/source/evoke.cc:
```diff
@@ -520,13 +522,64 @@ bool skill_has_manual(skill_type skill)
+    crate.props.erase(ARCHAEOLOGIST_CRATE_ITEM);
+}
+
+void archaeologist_read_tome(item_def& tome)
```
call update_can_train() here?

NormalPerson7 commented on crawl-ref/source/artefact.cc:
```diff
@@ -303,6 +303,11 @@ bool is_special_unrandom_artefact(const item_def &item)
+{
+    return !(get_unrand_entry(item.unrand_idx)->flags & UNRAND_FLAG_NOGEN);
+}
+
```
Now that I think of it, this has been obsoleted by `archaeologist_unrands()` which already checks for the NOGEN flag, so we can remove this.

NormalPerson7 committed:
Remove obsolete function can_generate(), add braces

Apparently when an `if` spans multiple lines, we're supposed to add
braces even when the following sequence is one line long.

NormalPerson7 committed:
Minor fixes to archaeologist

This commit does three things:
 - makes it so that the skill is immediately trainable and shows
   properly on the skill menu when the manual is unlocked at xl 3;
 - makes it so that the messages "you have an epiphany!" and "the crate
   breaks open revealing" don't occur every time the manual and unrand
   are picked up; and
 - adds a force-more to when the dusty tome is revealed and the crate
   opens.

NormalPerson7 commented:
Merged in four commits. Thanks for this pull request, it made everything a whole lot easier with this merge.
